URL-encode Location header in 201 response (#588)

league opened the pull request.

ruslantalpa commented on src/PostgREST/App.hs:
```diff
@@ -24,6 +24,7 @@ import           Text.ParserCombinators.Parsec (parse)
 
 import           Network.HTTP.Types.Header
 import           Network.HTTP.Types.Status
+import           Network.HTTP.Types.URI
```
maybe specify exactly the imported functions

ruslantalpa commented on test/Feature/InsertSpec.hs:
```diff
@@ -145,7 +145,7 @@ spec = do
             matchBody    = Just inserted
           , matchStatus  = 201
-          , matchHeaders = ["Location" <:> [str|/json?data=eq.{"foo":"bar"}|]]
+          , matchHeaders = ["Location" <:> "/json?data=eq.%7B%22foo%22%3A%22bar%22%7D"]
```
You mentioned somewhwere that there is a problem getting back the record with the value encoded, does it work now as it is?

league replied:
No, JSON keys aren't working for me, whether it's URL-encoded or not, and whether it's a list or an object. I'll add tests for those.

ruslantalpa commented:
Although this quick fix works for now, and since the location string is not calculated entirely on the db side maybe it would make sense to get back from sql an array of some sorts for the location (instead of a string) and have hasql decode it into haskell types and then later recombine them into an url.
That is if there is a significant speedup or correctness to worth the effort.

To summarise, if the returned location string can always be used to get back the record then this PR looks good.

ruslantalpa commented:
I just checked on your branch why the tests fail,
short version is that `json` type does not provide a equality operator ... i.e. it's useless as a primary key :)
the generated query is this
```
WITH pg_source AS (
        SELECT  "test"."json".* 
        FROM  "test"."json"  
        WHERE  "test"."json"."data" = '{"foo":"bar"}'::unknown  
        ) 
SELECT (SELECT pg_catalog.count(1) FROM  "test"."json" WHERE  "test"."json"."data" = '{"foo":"bar"}'::unknown ) AS total_result_set, pg_catalog.count(t) AS page_total, NULL::text[] AS header, coalesce(array_to_json(array_agg(row_to_json(t))), '[]')::character varying AS body
FROM ( SELECT * FROM pg_source LIMIT ALL OFFSET 0) t ",100,,"
```

which is correct.
I tried just changing the column to jsonb but that breaks other tests so i say we should just remove the additional tests that check you can get back the record by location and merge this PR (if everything looks good, i noticed there is a new commit and i did not look at it)

ruslantalpa commented:
Looked over the latest commit. Looks good.
I say remove the new failing part of the tests and it's good to merge

league commented:
Great. I rebased to current master and squashed some of the churn – hope that worked okay, I'm not an expert yet on git-rebase! 0a5ddc4 is the one we want, but the PR is showing other commits from master too, maybe that's normal.

begriffs commented:
@league try this

```bash
git remote add upstream https://github.com/begriffs/postgrest.git
git fetch upstream
git rebase upstream/master
```

(If that doesn't work we can try a `reset --hard` and a `cherry-pick`, but let's see)

ruslantalpa commented:
For rebase add -i parameter, it will give you the chance to squash all commits into one

ruslantalpa commented:
Oh, you already did that :)

league commented:
Yeah, I think this is clean now, and squashed to just the one commit. The other 5 it's showing are already in master. Not sure if there was a way to fetch upstream and rebase that would have prevented that.

begriffs commented:
OK here's a way to fix up your branch

```bash
git remote add upstream https://github.com/begriffs/postgrest.git
git fetch upstream
git reset --hard upstream/master
git cherry-pick 0a5ddc4
git push -f origin master
```

league commented:
Ah, that's more like it. I'm not sure what went wrong the first time. Possibly I did a pull instead of fetch from upstream, before doing the rebase -i.

ruslantalpa commented:
you should add a CHANGELOG line

begriffs commented on src/PostgREST/QueryBuilder.hs:
```diff
@@ -61,9 +61,22 @@ import           Data.Scientific         ( FPFormat (..)
+    is represented as a list of strings containing variable bindings like
+    @"k1=eq.42"@. If unused, it's null/Nothing rather than the empty list
+    because 'PostgreSQL.Binary.Decoder.arrayDimension' cannot decode an empty
+    array!
```
Very helpful comment.

Do you know if this is a bug in `arrayDimension`? It looks like an intentional library decision: https://github.com/nikita-volkov/postgresql-binary/blob/da82acf62855012ddecdcd6378d8fba69ca77b31/library/PostgreSQL/Binary/Decoder.hs#L499-L502

@nikita-volkov can you give us more context about this function?

nikita-volkov replied:
Can please provide some context of the issue?

league replied:
Hi @nikita-volkov, we were doing a `SELECT array_agg(...)` and decoding it into `[ByteString]`. But sometimes the query will produce zero rows, so the aggregator function should return an empty array `{}::text[]`. I expected that to decode to the empty list, but instead it produces `failure "A missing dimension length"` in `PostgreSQL.Binary.Decoder.arrayDimension`.

In the commit shown here, we just worked around it by decoding to `Maybe [ByteString]` instead, and that works fine.

We're using `Hasql.Decoders` (`HD`) which in turn calls `PostgreSQL.Binary.Decoder`. Something like this:

```haskell
-- works, producing Maybe [ByteString]:
HD.nullableValue $ HD.array $ HD.arrayDimension replicateM $ HD.arrayValue HD.bytea
-- fails with "A missing dimension length" when aggregated result set is empty:
HD.value $ HD.array $ HD.arrayDimension replicateM $ HD.arrayValue HD.bytea
```

nikita-volkov replied:
Okay. Thanks. I'll look into that and post back.

nikita-volkov replied:
It seems like you just need to upgrade your Hasql dependency bounds, because the latest releases have included [a fix for that](https://github.com/nikita-volkov/hasql/issues/46).

begriffs replied:
@nikita-volkov perfect!

@league I'll make a PR to update all deps to their newest version including hasql stuff. Once that's merged could you send a PR to refactor this part of the code?

league replied:
@begriffs Yeah, ping me from that dependency PR, if you remember.

begriffs commented on src/PostgREST/App.hs:
```diff
@@ -235,6 +235,14 @@ app dbStructure conf apiRequest =
 
+splitKeyValue :: BS.ByteString -> (BS.ByteString, BS.ByteString)
+splitKeyValue kv = (k, BS.tail v)
+  where (k, v) = BS.break (== '=') kv
```
Initially I worried about `=` signs in the value itself but it's fine, since `break` splits around the first one only. Good to know about this function, I hadn't noticed it in base.

begriffs commented:
@league thanks for this work, it's a significant improvement for inserts. Also really like your Haskell coding style. I'll merge this as is and we can improve the situation with `arrayDimension` later if the postgresql-binary package changes.
